### PR TITLE
Add trusted CA bundle support for OpenShift

### PR DESCRIPTION
**Pull request by objectiser**

### CODE REVIEW

**Comment by objectiser** [comment]
@kevinearls Would it be possible to kick off a test of this PR on OCP? 4.4 and 4.2 would be good. In addition to any automated tests, can you just check the Jaeger UI is available via the route.

**Comment by kevinearls** [comment]
@objectiser When I ran the CI tests on OCP 4.4 I got two failures.

The operator log contained a number of these messages: https://www.google.com/url?q=https://primetime.bluejeans.com/a2m/live-event/pahjkwua&sa=D&ust=1591558608965000&usg=AOvVaw3tHOqaVaXzpqQt1V4l6Iq1

The tests create a Jaeger instance for each test case, so maybe we need to delete the configmap when the jaeger instance is deleted.

**Comment by objectiser** [comment]
@kevinearls I've updated the PR, as its possible a user would have already configured the volumeMount with a trusted CA bundle (i.e. Service Mesh will be doing that).

I tried to manually deploy to reproduce the problem you were experiencing, and the configmap seems to be correctly deleted when the Jaeger instance CR is removed. Would you be able to retest - and if necessary update the tests to wait for the configmap to be undeployed (if that is the issue)?

**Comment by kevinearls** [comment]
@objectiser The ES index tests are still failing on me, although I found out that most of the errors in the log occurred immediately after the Jaeger instance was created: https://github.com/jaegertracing/jaeger-operator/blob/master/test/e2e/elasticsearch_test.go#L140-L146

That doesn't make much difference initially, as the smoke test step passes.  However, the cron job that is supposed to get created when we enable the index cleaner never appears.  https://github.com/jaegertracing/jaeger-operator/blob/master/test/e2e/elasticsearch_test.go#L256-L265. I don't see anything in the operator log about this.  I will try to get more info tomorrow if possible.

**Comment by objectiser** [comment]
@kevinearls Would it also be worth trying the same tests against `jaegertracing/jaeger-operator:master`, to see if the same tests are failing, or whether it does appear to be due to the PR?

**Comment by kevinearls** [comment]
@objectiser It's definitely due to the PR, tests pass when I run against master.

**Comment by jpkrohling** [comment]
Is this related to #1043?

**Comment by objectiser** [comment]
@jpkrohling No this relates to TRACING-1208 - first attempt to implement was just mounting the voume on the oauth proxy container, which is the original problem - but when that failed in OCP I thought it may need the volume mounted in all components (which is also the same solution that Service Mesh would be implementing by adding the volume/mounts in their Jaeger templates).

So we need to find a solution to TRACING-1208 and decide whether volume/mounts required across all components or just for oauth-proxy. Ideally we need the problem reproduced in a OCP cluster to properly test with custom CA bundle.

**Comment by jpkrohling** [comment]
Running this PR locally, I noticed that the volume/volumemount is added to the query container, not to the OAuth container:

```yaml
spec:
  containers:
  - args:
    - --collector.grpc.tls.cert=/etc/tls-config/tls.crt
    - --collector.grpc.tls.enabled=true
    - --collector.grpc.tls.key=/etc/tls-config/tls.key
    - --query.ui-config=/etc/config/ui.json
    - --reporter.grpc.tls.ca=/var/run/secrets/kubernetes.io/serviceaccount/service-ca.crt
    - --reporter.grpc.tls.enabled=true
    - --reporter.grpc.tls.server-name=simplest-collector-headless.default.svc.cluster.local
    - --reporter.type=grpc
    - --sampling.strategies-file=/etc/jaeger/sampling/sampling.json
    env:
    - name: SPAN_STORAGE_TYPE
      value: memory
    - name: COLLECTOR_ZIPKIN_HTTP_PORT
      value: "9411"
    image: jaegertracing/all-in-one:1.18.0
    imagePullPolicy: IfNotPresent
    livenessProbe:
      failureThreshold: 5
      httpGet:
        path: /
        port: 14269
        scheme: HTTP
      initialDelaySeconds: 5
      periodSeconds: 15
      successThreshold: 1
      timeoutSeconds: 1
    name: jaeger
    ports:
    - containerPort: 5775
      name: zk-compact-trft
      protocol: UDP
    - containerPort: 5778
      name: config-rest
      protocol: TCP
    - containerPort: 6831
      name: jg-compact-trft
      protocol: UDP
    - containerPort: 6832
      name: jg-binary-trft
      protocol: UDP
    - containerPort: 9411
      name: zipkin
      protocol: TCP
    - containerPort: 14267
      name: c-tchan-trft
      protocol: TCP
    - containerPort: 14268
      name: c-binary-trft
      protocol: TCP
    - containerPort: 16686
      name: query
      protocol: TCP
    - containerPort: 14269
      name: admin-http
      protocol: TCP
    - containerPort: 14250
      name: grpc
      protocol: TCP
    readinessProbe:
      failureThreshold: 3
      httpGet:
        path: /
        port: 14269
        scheme: HTTP
      initialDelaySeconds: 1
      periodSeconds: 10
      successThreshold: 1
      timeoutSeconds: 1
    resources: {}
    terminationMessagePath: /dev/termination-log
    terminationMessagePolicy: File
    volumeMounts:
    - mountPath: /etc/config
      name: simplest-ui-configuration-volume
      readOnly: true
    - mountPath: /etc/jaeger/sampling
      name: simplest-sampling-configuration-volume
      readOnly: true
    - mountPath: /etc/tls-config
      name: simplest-collector-tls-config-volume
      readOnly: true
    - mountPath: /etc/pki/ca-trust/extracted/pem
      name: simplest-trusted-ca
      readOnly: true
    - mountPath: /var/run/secrets/kubernetes.io/serviceaccount
      name: simplest-ui-proxy-token-bkm5j
      readOnly: true
  - args:
    - --cookie-secret=7+6bwAmtQOK6z5otl0X3RQ==
    - --https-address=:8443
    - --openshift-service-account=simplest-ui-proxy
    - --provider=openshift
    - --tls-cert=/etc/tls/private/tls.crt
    - --tls-key=/etc/tls/private/tls.key
    - --upstream=http://localhost:16686
    image: openshift/oauth-proxy:latest
    imagePullPolicy: Always
    name: oauth-proxy
    ports:
    - containerPort: 8443
      name: public
      protocol: TCP
    resources: {}
    terminationMessagePath: /dev/termination-log
    terminationMessagePolicy: File
    volumeMounts:
    - mountPath: /etc/tls/private
      name: simplest-ui-oauth-proxy-tls
    - mountPath: /var/run/secrets/kubernetes.io/serviceaccount
      name: simplest-ui-proxy-token-bkm5j
      readOnly: true
```

Your description seems to indicate that it should be added to the OAuth Proxy (which is also what makes sense to me). Once @jkandasa and/or @kevinearls are able to reproduce this in a test cluster, I'll update this PR to get the volume/volumemount added to the OAuth sidecar.

**Comment by objectiser** [comment]
@jpkrohling Must have made the mistake when changing from oauth proxy only to all components. However might be worth testing this PR yourself on OCP to see if you get the es-index-cleaner issue found by @kevinearls?

**Comment by jpkrohling** [comment]
The last state of this PR is passing the e2e tests in OpenShift 4.3:

```
Formatting code...
Building...
STEP 1: FROM registry.access.redhat.com/ubi8/ubi
STEP 2: ENV OPERATOR=/usr/local/bin/jaeger-operator     USER_UID=1001     USER_NAME=jaeger-operator
--> Using cache 935e0bf5a64a9bde337a055524dfbe0ade13e52cbd4810b2513c0ac8e7d7ee95
STEP 3: RUN INSTALL_PKGS="       openssl       " &&     yum install -y $INSTALL_PKGS &&     rpm -V $INSTALL_PKGS &&     yum clean all &&     mkdir /tmp/_working_dir &&     chmod og+w /tmp/_working_dir
--> Using cache 14e912f6e6bb423c86ff5e17644c6c5cd315bc4bb736d253aa8b10101095310a
STEP 4: COPY scripts/* /scripts/
--> Using cache 23e146deae55de9926c332aee4d2646a56b0ac3b83c29c2eaad6f685927e39d8
STEP 5: COPY build/_output/bin/jaeger-operator ${OPERATOR}
--> 0259cdd7b7e
STEP 6: ENTRYPOINT ["/usr/local/bin/jaeger-operator"]
--> 7a33f95ac29
STEP 7: USER ${USER_UID}
STEP 8: COMMIT quay.io/jpkroehling/jaeger-operator:latest
--> 446295a9a52
446295a9a5240a01d5c7ca22c8f49775bc76ea55df237a960cd11fac4312aff3
Pushing image quay.io/jpkroehling/jaeger-operator:latest...
Getting image source signatures
Copying blob f83864b38722 done  
Copying blob d777138324bc skipped: already exists  
Copying blob 92b864bfcfaa skipped: already exists  
Copying blob 808aa963fe95 skipped: already exists  
Copying blob 92af68d64d0b skipped: already exists  
Copying config 446295a9a5 done  
Writing manifest to image destination
Copying config 446295a9a5 [--------------------------------------] 0.0b / 3.4KiB
Writing manifest to image destination
Writing manifest to image destination
Storing signatures
Running Smoke end-to-end tests...
BUILD_IMAGE="quay.io/jpkroehling/jaeger-operator:latest" go test -tags=smoke ./test/e2e/...  -kubeconfig /mnt/storage/jpkroehling/Tools/openshift/tracing-1208-2/auth/kubeconfig -namespacedMan ../../deploy/test/namespace-manifests.yaml -globalMan ../../deploy/test/global-manifests.yaml -root .
ok  	github.com/jaegertracing/jaeger-operator/test/e2e	562.092s
Creating namespace default
Running Cassandra end-to-end tests...
ok  	github.com/jaegertracing/jaeger-operator/test/e2e	200.531s
Running Elasticsearch end-to-end tests...
ok  	github.com/jaegertracing/jaeger-operator/test/e2e	341.471s
# Elasticsearch requires labeled nodes. These labels are by default present in OCP 4.2
node/ip-10-0-140-154.ec2.internal not labeled
node/ip-10-0-141-223.ec2.internal not labeled
node/ip-10-0-146-160.ec2.internal not labeled
node/ip-10-0-151-17.ec2.internal not labeled
node/ip-10-0-172-192.ec2.internal not labeled
node/ip-10-0-172-250.ec2.internal not labeled
# This is not required in OCP 4.1. The node tuning operator configures the property automatically
# when label tuned.openshift.io/elasticsearch=true label is present on the ES pod. The label
# is configured by ES operator.
customresourcedefinition.apiextensions.k8s.io/prometheusrules.monitoring.coreos.com configured
customresourcedefinition.apiextensions.k8s.io/servicemonitors.monitoring.coreos.com configured
serviceaccount/elasticsearch-operator unchanged
clusterrole.rbac.authorization.k8s.io/elasticsearch-operator unchanged
clusterrolebinding.rbac.authorization.k8s.io/elasticsearch-operator-rolebinding unchanged
customresourcedefinition.apiextensions.k8s.io/elasticsearches.logging.openshift.io unchanged
deployment.apps/elasticsearch-operator configured
deployment.extensions/elasticsearch-operator image updated
Running Self provisioned Elasticsearch end-to-end tests...
ok  	github.com/jaegertracing/jaeger-operator/test/e2e	327.169s
Creating namespace kafka
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   647  100   647    0     0   1767      0 --:--:-- --:--:-- --:--:--  1767
100  302k  100  302k    0     0   227k      0  0:00:01  0:00:01 --:--:--  781k

customresourcedefinition.apiextensions.k8s.io/kafkas.kafka.strimzi.io unchanged
rolebinding.rbac.authorization.k8s.io/strimzi-cluster-operator-entity-operator-delegation unchanged
clusterrolebinding.rbac.authorization.k8s.io/strimzi-cluster-operator unchanged
rolebinding.rbac.authorization.k8s.io/strimzi-cluster-operator-topic-operator-delegation unchanged
customresourcedefinition.apiextensions.k8s.io/kafkausers.kafka.strimzi.io unchanged
clusterrole.rbac.authorization.k8s.io/strimzi-entity-operator unchanged
clusterrole.rbac.authorization.k8s.io/strimzi-cluster-operator-global unchanged
clusterrolebinding.rbac.authorization.k8s.io/strimzi-cluster-operator-kafka-broker-delegation unchanged
rolebinding.rbac.authorization.k8s.io/strimzi-cluster-operator unchanged
clusterrole.rbac.authorization.k8s.io/strimzi-cluster-operator-namespaced unchanged
clusterrole.rbac.authorization.k8s.io/strimzi-topic-operator unchanged
serviceaccount/strimzi-cluster-operator unchanged
clusterrole.rbac.authorization.k8s.io/strimzi-kafka-broker unchanged
customresourcedefinition.apiextensions.k8s.io/kafkatopics.kafka.strimzi.io unchanged
customresourcedefinition.apiextensions.k8s.io/kafkabridges.kafka.strimzi.io unchanged
customresourcedefinition.apiextensions.k8s.io/kafkaconnectors.kafka.strimzi.io unchanged
customresourcedefinition.apiextensions.k8s.io/kafkaconnects2is.kafka.strimzi.io unchanged
customresourcedefinition.apiextensions.k8s.io/kafkaconnects.kafka.strimzi.io unchanged
customresourcedefinition.apiextensions.k8s.io/kafkamirrormakers.kafka.strimzi.io unchanged
The Deployment "strimzi-cluster-operator" is invalid: spec.template.spec.containers[0].env[0].valueFrom: Invalid value: "": may not be specified when `value` is not empty
Creating namespace kafka
kubectl create namespace "kafka" 2>&1 | grep -v "already exists" || true
curl --fail --location "https://raw.githubusercontent.com/strimzi/strimzi-kafka-operator/0.16.2/examples/kafka/kafka-persistent-single.yaml" --output deploy/test/kafka-example.yaml --create-dirs
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   669  100   669    0     0   1911      0 --:--:-- --:--:-- --:--:--  1911

sed -i 's/size: 100Gi/size: 10Gi/g' deploy/test/kafka-example.yaml
kubectl -n "kafka" apply --dry-run=client -f deploy/test/kafka-example.yaml
kafka.kafka.strimzi.io/my-cluster configured (dry run)
kafka.kafka.strimzi.io/my-cluster unchanged
Running Streaming end-to-end tests...
ok  	github.com/jaegertracing/jaeger-operator/test/e2e	313.817s
Running Example end-to-end tests part 1...
ok  	github.com/jaegertracing/jaeger-operator/test/e2e	189.879s
Running Example end-to-end tests part 2...
ok  	github.com/jaegertracing/jaeger-operator/test/e2e	244.268s
Running OpenShift Example end-to-end tests...
ok  	github.com/jaegertracing/jaeger-operator/test/e2e	125.327s
Running generate end-to-end tests...
ok  	github.com/jaegertracing/jaeger-operator/test/e2e	38.680s
```

**Comment by objectiser** [comment]
@jpkrohling If you are happy with the PR as is, can you approve and merge? If any other issues are found in subsequent QE testing they can be fixed in a followup PR.

**Comment by jkandasa** [comment]
@objectiser @jpkrohling I verified this PR via `quay.io/jpkroehling/jaeger-operator:TRACING-1208`.

LGTM :+1:

File: pkg/config/ca/ca_test.go [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+	"k8s.io/apimachinery/pkg/types"
+
+	v1 "github.com/jaegertracing/jaeger-operator/pkg/apis/jaegertracing/v1"
+)
+
+func TestGetWithoutTrustedCA(t *testing.T) {
+	viper.Set("platform", "other")
+	defer viper.Reset()
+
+	jaeger := v1.NewJaeger(types.NamespacedName{Name: "TestGetWithoutTrustedCA"})
```
Comment (jpkrohling) [inline pkg/config/ca/ca_test.go:18]: I think we started using a generic name some time ago. Was it changed recently, to use individual names for the instances in each test? If not, there's no need to change this now.

File: pkg/config/ca/ca.go [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+			OwnerReferences: []metav1.OwnerReference{{
+				APIVersion: jaeger.APIVersion,
+				Kind:       jaeger.Kind,
+				Name:       jaeger.Name,
+				UID:        jaeger.UID,
+				Controller: &trueVar,
+			}},
+		},
+		Data: map[string]string{
+			"ca-bundle.crt": "",
```
Comment (jpkrohling) [inline pkg/config/ca/ca.go:53]: Tests are showing that this is fine, but I wonder if there's a reason to have this as empty first? Also, would the reconciliation phase revert this to empty? If so, OpenShift seems to be refilling this quite fast (possibly with a webhook, before changes are actually applied).

Comment (chlunde) [reply pkg/config/ca/ca.go:53]: might be an idea to check with `kubectl get -w configmap -o yaml` - could be a potencial race?

**Pull request merged**